pkg/clusterid/parse_cluster_id.go

```diff
@@ -5,13 +5,59 @@ import (
 
 	"github.com/pkg/errors"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/buildinfo"
+	"github.com/stackrox/rox/pkg/logging"
 	"github.com/stackrox/rox/pkg/mtls"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
+var (
+	once     sync.Once
+	instance *parserWrapper
+	log      = logging.LoggerForModule()
+)
+
+// Parser defines an interface with the function to parse the cluster ID from the service cert.
+type Parser interface {
+	ParseClusterIDFromServiceCert(expectedServiceType storage.ServiceType) (string, error)
+}
+
+// parserWrapper is a singleton that wraps the Parser interface.
+// This allows us to override the implementation of the Parser interface for testing purposes.
+// This override mechanism is protected in ReleaseBuilds, and it will only work for DevelopmentBuilds.
+type parserWrapper struct {
+	parser Parser
+}
+
+// parserImpl is the implementation of the Parser interface.
+type parserImpl struct {
+}
+
+// GetParser returns the parserWrapper singleton.
+func GetParser() *parserWrapper {
+	once.Do(func() {
+		instance = &parserWrapper{
+			parser: &parserImpl{},
+		}
+	})
+	return instance
+}
+
+// Override the internal parser. This should only be used for testing.
+func (p *parserWrapper) Override(parser Parser) {
```

Comment:
If you follow the suggestion above, then you don't need the receiver pointer. 

And the overwrite would be `p = parser`

Reply:
🤔 I think it’s not possible to have `Override` as a member function of `parserImpl` since it receives the `Parser` interface and you cannot (or I don’t know how to) assign `p` of type `parserImpl` to a variable of type `Parser`.

Reply:
The point was about making `Override` static rather than a member function of `parserImpl`. I'm fine if you want to keep it this way. But I personally still haven't seen a reason why the extra wrapper layer is necessary (or even beneficial).

Reply:
Gotcha! I've changed the code to remove the wrappers in [e2f16f8](https://github.com/stackrox/stackrox/pull/6541/commits/e2f16f8dd6bb588ebb39e106e55921e6c3473d14)